fix(api): add supported wavelengths to runtime error when initializing the plate reader.

vegano1 opened the pull request:
# Overview

Add the supported wavelengths to the runtime error raised when an invalid wavelength is provided.

Closes: [RQA-3568](https://opentrons.atlassian.net/browse/RQA-3568)

## Test Plan and Hands on Testing

- [x] Make sure the supported wavelengths are reported in the runtime error

## Changelog

- add `InvalidWavelengthError` to denote incompatible wavelengths with the current plate reader.
- use the `InvalidWavelengthError` and format a message to clarify what wavelengths can be used.

## Review requests

- I'm open to word smith's suggestions

## Risk assessment

low, log only change

[RQA-3568]: https://opentrons.atlassian.net/browse/RQA-3568?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
fix(api): add supported wavelengths to runtime error when initializing the plate reader.

ecormany commented on api/src/opentrons/protocol_engine/commands/absorbance_reader/initialize.py:
```diff
@@ -70,7 +70,10 @@ async def execute(self, params: InitializeParams) -> SuccessData[InitializeResul
             if unsupported_wavelengths:
-                raise ValueError(f"Unsupported wavelengths: {unsupported_wavelengths}")
+                raise ValueError(
+                    f"Unsupported wavelengths: {unsupported_wavelengths}, "
+                    f" use {supported_wavelengths} instead."
```
Small change to make this a bit more like the error we have for loading an invalid module, which I like the format of:

```suggestion
                    f"Unsupported wavelengths: {unsupported_wavelengths}. "
                    f" Use one of {supported_wavelengths} instead."
```

ecormany reviewed (approved):
Nice, so simple! Would love to be able to surface this in analysis, but that's for another day…

vegano1 committed:
Update api/src/opentrons/protocol_engine/commands/absorbance_reader/initialize.py

Co-authored-by: Ed Cormany <edward.cormany@opentrons.com>

sfoster1 commented on api/src/opentrons/protocol_engine/commands/absorbance_reader/initialize.py:
```diff
@@ -70,7 +70,10 @@ async def execute(self, params: InitializeParams) -> SuccessData[InitializeResul
             )
             if unsupported_wavelengths:
-                raise ValueError(f"Unsupported wavelengths: {unsupported_wavelengths}")
+                raise ValueError(
```
This should be an error from `protocol_engine.errors` that gets associated with an error code - maybe invalid protocol data, but ok to fall back to 4000 general error instead. If it's just a raw exception it will get a scarier UI.

Also this seems like it's just dumping a set to string which will look like this: "Unsupported wavelengths: {1000, 1400, 1200}. Use one of {500, 800, 1000} instead." Feels a little terse, maybe make it `{', '.join([str(wavelength) + 'nm' for wavelength in unsupported_wavelengths])` instead?

SyntaxColoring reviewed (approved):
All my homies love actionability.

sfoster1 reviewed (changes requested):
Comments inline, but should use a protocol engine error and maybe massage the formatting of the wavelengths.

vegano1 committed:
add ProtocolEngine error InvalidWavelengthError

vegano1 committed:
Merge branch 'add-supported-wavelengths-to-runtime-error' of https://github.com/Opentrons/opentrons into add-supported-wavelengths-to-runtime-error

vegano1 commented:
@sfoster1 @ecormany 

Added `InvalidWavelengthError` and fixed up the formatting, here's what that looks like.

<img width="760" alt="Screen Shot 2024-11-14 at 9 13 13 AM" src="https://github.com/user-attachments/assets/b3cb8c61-4536-4195-b591-5ea878022853">

sfoster1 reviewed (approved):
Excellent, thank you!
